pkg/query-service/rules/thresholdRule.go

```diff
@@ -355,21 +354,37 @@ func (r *ThresholdRule) CheckCondition(v float64) bool {
 func (r *ThresholdRule) prepareQueryRange(ts time.Time) *v3.QueryRangeParamsV3 {
 	// todo(amol): add 30 seconds to evalWindow for rate calc
 
+	// todo(srikanthccv): make this configurable
+	// 2 minutes is reasonable time to wait for data to be available
+	// 60 seconds (SDK) + 10 seconds (batch) + rest for n/w + serialization + write to disk etc..
+	start := ts.Add(-time.Duration(r.evalWindow)).UnixMilli() - 2*60*1000
```

Comment:
Not sure about the intent here.  [is this to solve rate calculation problem?]

wouldn't this add 2 min delay in raising alerts?  The alert manager also groups alerts in 5(?) mins window.  So the actual notification would get pushed out a further. 

If this is for a specific use case, we could introduce a command line parameter similar to --groupInterval  or --sendDelay etc. I see a todo there but may be it can be prioritised in case things dont work out, customers can rollback.

Reply:
This is not a specific use case or rate but a problem in general. The duration we consider for the alerts evaluation is adjusted so that the data is still in motion is considered. Assume the alert evaluator sets the end time as 3:19:01 pm; in practice, the data for this interval is yet to arrive or be available in ClickHouse. Currently, when we create a query with `timestamp between start and end`; it triggers false alerts because of incomplete data, which is more annoying to users.

Reply:
Discussed in Slack